### ISSUE TITLE
Refactoring Mailu's frontend

### ISSUE THREAD

**Issue by kaiyou** [Mailu/Mailu#272]
After moving the HTTP frontend to `traefik`, multiple users pointed out that `nginx` could have been used as a mail reverse proxy. I had not thought of that kind of architecture yet, but it did sound interesting enough so that I stopped my work on `traefik` while reading through the documentation.

Nginx as a mail proxy has multiple drawbacks:
- it does not support disabling authentication to IMAP, which means password is still shipped to Dovecot
- it does not support TLS to the backend, which leaves securing the channel between hosts up to the admin (it's already the case for webmails, but still)
- the documentation is sparse because it is not used that much except for large e-mail infrastructures.

However, advantages are non-negligeable:
- authentication is proxified to an HTTP endpoint, which leaves us with an interesting degree of freedom
- we could implement authentication rate limits and account lockouts
- we could properly limit features (imap, smtp, source ip address, etc.) per account
- we could implement application passwords to avoid leaking account login and password in third party applications
- we could then try and enforce 2FA for Web access and actually have increased security by disabling password authentication (except for application passwords) on IMAP/SMTP
- we could simplify Postfix/Dovecot confiurations by moving all the authentication logics to a Web API
- we could do proper letsencrypt on the frontend and have it terminate all TLS connections, with a very uniform TLS configuration
- we would not expose a container with access to email databases before the user is properly authenticated, which limits attack surface

I started playing around with Nginx and I believe this is a plausible solution. I will start a new branch and see how fast this can be implemented. In the meanwhile, I added a disclaimer on the main README to discourage people from using `master` without first checking the configuration, as this could have disastrous consequences if we move from authenticating in the SMTP and IMAP container to authenticating in the frontend.

Finally, this would mean that Mailu could only support a single long term frontend as the adherence to Nginx would be increased. I will try and keep the `TLS_FLAVOUR` option for people who are currently using `nginx-no-https`.

**Comment by kaiyou** [comment]
I am setting this as planned for `1.5` as motivator. Traefik will remain a transitional solution if we do not succeed by mid-october, as it provies a proper letsencrypt implementation.

**Comment by kaiyou** [comment]
A first version was just pushed to branch `refactor-nginx`. It does not support TLS for IMAP and SMTP yet. Also, security was not fully tested, so be very careful.

**Comment by SunMar** [comment]
Nice work! :)

I have a small suggestion for `nginx.conf`, Nginx by default does redirects (for example if you access /admin it redirects to /admin/ because it's a directory) with a full URL (including scheme and port). If you use Nginx without TLS and put your own reverse-proxy or something in front of it that handles HTTPS or is on a different port or even a different host, that can cause issues because Nginx will redirect to a wrong URL. An easy solution is to add [`absolute_redirect off;`](http://nginx.org/en/docs/http/ngx_http_core_module.html#absolute_redirect) in `nginx.conf` so Nginx does relative redirects and you won't get redirected to the wrong scheme or port because that's handled browser side. This directive wasn't added until 1.11.8 but if I look in the Alpine package repository for the `edge` channel (which I believe you are using as a base for the Docker container now) I see Nginx 1.12 so that should be fine.

**Comment by SunMar** [comment]
On that same note I would also suggest changing other redirects in `nginx.conf` to relative redirects where possible.

https://github.com/Mailu/Mailu/blob/ba4ed579df94743460706f99017716a81ce23fe9/nginx/conf/nginx.conf#L50

That could be changed to `return 301 /webmail/;` and the same would apply to other `return` usages in the configuration. For the end user it should not matter and it prevents issues in more exotic setups.

**Comment by kaiyou** [comment]
Indeed nginx runs on 1.12

```
nginx version: nginx/1.12.1
built with LibreSSL 2.5.5
TLS SNI support enabled
```

I will apply these, thank you for the suggestion :)

**Comment by kaiyou** [comment]
Has anyone else tested this on a proper server during the week? I will set it up on `test.mailu.io` and ask for some security tests before we can consider merging this work.

**Comment by SaiGonSeamus** [comment]
We have two test servers running here now. One in a DMZ on in the cloud. Tell us which branches / versions we should put up on them and we'll test all you need. Glad to be able to contribute in some way to the project.

**Comment by kaiyou** [comment]
The `nginx-refactor` branch should already be built by Docker Hub. Also, make sure you use the new `docker-compose.yml` and `.env` files from that branch :)

**Comment by kaiyou** [comment]
This week-end, I will to start a security test. Then add features like brute-force protection per source-ip, filtering features upon connection, and maybe application passwords per user.

**Comment by kaiyou** [comment]
This has been running on my test server for some time now, with now specific issue. The plan is to merge into master, delay the release for a couple days in order to integrate it directly.

**Comment by curry684** [comment]
Does this mean 1.5 is going to happen soon and have a clean upgrade path?

**Comment by kaiyou** [comment]
It is going to happen soon (I would say 4th of november instead of last week). The upgrade path will be documented but might be hurtful because both `docker-compose.yml` and `.env` changed.

**Comment by curry684** [comment]
Ok, in that case I'll delay a pending upgrade I have here as I need the domain aliases feature  😉 

If only the 2 external files changed I think we can consider the upgrade 'clean', it would be more worrying if for example the DB needed conversions.

**Comment by kaiyou** [comment]
Nope, all database changes are based on Alembic migrations, so conversion is done automatically at startup based on current and target database version.

**Comment by kaiyou** [comment]
This is now merged into master, after a couple days of intensive testing and bug fixing.
